### PR TITLE
Refactor parfor: better exception from passes

### PR DESCRIPTION
**Pull request by sklam**
as titled.

### CODE REVIEW

File: numba/core/errors.py [hunk 1]
```diff
@@ -493,6 +493,12 @@ class UnsupportedError(NumbaError):
     pass
 
 
+class RewriteUnsupportedError(UnsupportedError):
```
Comment (stuartarchibald) [inline numba/core/errors.py]: Wondering if `UnsupportedRewriteError` sounds better?

File: numba/parfors/array_analysis.py [hunk 2]
```diff
@@ -1766,11 +1769,11 @@ def _analyze_op_call(self, scope, equiv_set, expr):
                 expr.args = args[1:]
             return result
 
-    def _analyze_op_call___builtin___len(self, scope, equiv_set, args, kws):
+    def _analyze_op_call___builtin___len(self, scope, equiv_set, loc, args, kws):
         # python 2 version of len()
```
Comment (stuartarchibald) [inline numba/parfors/array_analysis.py]: Py 2... dead?

File: numba/parfors/parfor.py [hunk 3]
```diff
@@ -1836,7 +1838,9 @@ def _numpy_to_parfor(self, equiv_set, lhs, expr):
         if call_name in ['zeros', 'ones'] or mod_name == 'numpy.random':
             return self._numpy_map_to_parfor(equiv_set, call_name, lhs, args, kws, expr)
         # return error if we couldn't handle it (avoid rewrite infinite loop)
-        raise NotImplementedError("parfor translation failed for ", expr)
+        raise errors.RewriteUnsupportedError(
+            f"parfor translation failed for {expr}", loc=exprloc,
```
Comment (stuartarchibald) [inline numba/parfors/parfor.py]: Should that be `expr.loc`? If so, no build failures suggests this isn't tested?

Comment (sklam) [reply numba/parfors/parfor.py]: good catch. some of the exception raise in Parfor are not covered by the testsuite

Comment (sklam) [reply numba/parfors/parfor.py]: not sure how to trigger this one.

File: numba/core/errors.py (hunk 1 above)
Comment (sklam) [reply numba/core/errors.py]: it does

File: numba/parfors/array_analysis.py (hunk 2 above)
Comment (sklam) [reply numba/parfors/array_analysis.py]: i'll go after all the `__builtin__` references

Comment (sklam) [reply numba/parfors/array_analysis.py]: they are removed

File: numba/core/errors.py (hunk 1 above)
Comment (sklam) [reply numba/core/errors.py]: renamed

**Pull request merged**